File: app/parser.py
\
import re
from bs4 import BeautifulSoup
from datetime import datetime
from typing import List, Dict

# Regexes to capture JS new Date(...) and event blocks
DATE_RE = re.compile(r"new Date\(\s*(\d{4})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*\)")
EVENT_BLOCK_RE = re.compile(
    r"\{[^{}]*?['\"]start['\"]\s*:\s*new Date\([^\)]*\)[\s\S]*?['\"]end['\"]\s*:\s*new Date\([^\)]*\)[\s\S]*?\}",
    re.MULTILINE
)

def _parse_js_date(match: re.Match) -> datetime:
    """Convert JS new Date(Y, M, D, H, m) to Python datetime (note: JS month is 0-based)."""
    y, mth, d, hh, mm = map(int, match.groups())
    return datetime(y, mth + 1, d, hh, mm)

def _strip_html(html: str) -> str:
    """Strip HTML to plain text, collapsing whitespace."""
    soup = BeautifulSoup(html or "", "html.parser")
    return " ".join(soup.get_text(" ").split())

def extract_events(js_blob: str) -> List[Dict]:
    """
    Parse the JS blob that contains something like:

      <script> ... events: [{
        start: new Date(2025, 8, 18, 18, 10),
        end: new Date(2025, 8, 18, 19, 50),
        title: '<table>...',
        body: '<table>...'
      }, ...] </script>

    Returns a list of normalized dicts:
      { 'start': datetime, 'end': datetime, 'summary': str, 'location': str, 'description': str }
    """
    events = []
    if not js_blob:
        return events

    for block in EVENT_BLOCK_RE.findall(js_blob):
        # start
        sm = DATE_RE.search(block)
        if not sm:
            continue
        start = _parse_js_date(sm)

        # end (second occurrence)
        em = DATE_RE.search(block, sm.end())
        if not em:
            continue
        end = _parse_js_date(em)

        # Extract title/body (single-quoted or double-quoted)
        title_match = re.search(r"""['"]title['"]\s*:\s*(['"])(.*?)\1""", block, re.DOTALL)
        body_match  = re.search(r"""['"]body['"]\s*:\s*(['"])(.*?)\1""", block, re.DOTALL)

        def unescape_js(s: str) -> str:
            # handle common JS-escaped quotes and newlines
            return s.replace(r"\'", "'").replace(r'\"', '"').replace(r"\n", " ").replace(r"\r", " ")

        title_html = unescape_js(title_match.group(2)) if title_match else ""
        body_html  = unescape_js(body_match.group(2)) if body_match else ""

        title_txt = _strip_html(title_html)
        body_txt  = _strip_html(body_html)

        # Heuristic for a room/location
        location = ""
        sala = re.search(r"(Sala\s+[A-Za-z0-9\-]+|[A-Z]-\d{2,3})", title_txt)
        if sala:
            location = sala.group(0)

        events.append({
            "start": start,
            "end": end,
            "summary": title_txt[:200] if title_txt else "Class",
            "location": location,
            "description": body_txt[:2000] if body_txt else ""
        })
    return events

File: src/parser.ts
import * as cheerio from 'cheerio';
import { Event } from './types.js';

// Regex patterns for parsing JavaScript date objects and event blocks
const DATE_REGEX = /new Date\(\s*(\d{4})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*,\s*(\d{1,2})\s*\)/g;
const EVENT_BLOCK_REGEX = /\{[^{}]*?['"]start['"]\s*:\s*new Date\([^\)]*\)[\s\S]*?['"]end['"]\s*:\s*new Date\([^\)]*\)[\s\S]*?\}/g;

/**
 * Parse JavaScript Date constructor to native Date object
 * Note: JavaScript months are 0-based, so we add 1
 */
function parseJsDate(match: RegExpMatchArray): Date {
  if (!match || match.length < 6) {
    throw new Error('Invalid date match');
  }
  const [, year, month, day, hour, minute] = match.map(Number);
  return new Date(year, month, day, hour, minute);
}

/**
 * Strip HTML tags and normalize whitespace
 */
function stripHtml(html: string): string {
  if (!html) return '';
  const $ = cheerio.load(html);
  return $.text().replace(/\s+/g, ' ').trim();
}

/**
 * Unescape JavaScript string literals
 */
function unescapeJs(str: string): string {
  return str
    .replace(/\\'/g, "'")
    .replace(/\\"/g, '"')
    .replace(/\\n/g, ' ')
    .replace(/\\r/g, ' ')
    .replace(/\\t/g, ' ');
}

/**
 * Extract room/location from text using heuristics
 */
function extractLocation(text: string): string {
  const locationMatch = text.match(/(Sala\s+[A-Za-z0-9\-]+|[A-Z]-\d{2,3})/);
  return locationMatch ? locationMatch[0] : '';
}

/**
 * Parse JavaScript blob containing event data and extract calendar events
 * 
 * Expected format:
 * <script> ... events: [{
 *   start: new Date(2025, 8, 18, 18, 10),
 *   end: new Date(2025, 8, 18, 19, 50),
 *   title: '<table>...',
 *   body: '<table>...'
 * }, ...] </script>
 */
export function extractEvents(jsBlob: string): Event[] {
  if (!jsBlob) return [];

  const events: Event[] = [];
  const eventBlocks = jsBlob.match(EVENT_BLOCK_REGEX);

  if (!eventBlocks) return events;

  for (const block of eventBlocks) {
    try {
      // Extract start date
      const startMatches = Array.from(block.matchAll(DATE_REGEX));
      if (startMatches.length < 2) continue;

      const startDate = parseJsDate(startMatches[0]);
      const endDate = parseJsDate(startMatches[1]);

      // Extract title and body with proper quote handling
      const titleMatch = block.match(/['"]title['"]\s*:\s*(['"])(.*?)\1/s);
      const bodyMatch = block.match(/['"]body['"]\s*:\s*(['"])(.*?)\1/s);

      const titleHtml = titleMatch ? unescapeJs(titleMatch[2]) : '';
      const bodyHtml = bodyMatch ? unescapeJs(bodyMatch[2]) : '';

      const titleText = stripHtml(titleHtml);
      const bodyText = stripHtml(bodyHtml);
      const location = extractLocation(titleText);

      events.push({
        start: startDate,
        end: endDate,
        summary: titleText.slice(0, 200) || 'Class',
        location,
        description: bodyText.slice(0, 2000),
      });
    } catch (error) {
      console.warn('Failed to parse event block:', error);
      continue;
    }
  }

  return events;
}
